language-support/ts/daml-ledger/index.ts

```diff
@@ -493,7 +493,7 @@ class QueryStreamsManager {
 
             const ws = new WebSocket(manager.url, manager.protocols);
```

Comment:
If I'm counting that right, this block of code has 4 different variables called `ws` which need to be distinguished. I'm not surprised that resulted in a bug here.

There seems to be:
- Two `const ws` with nested scopes, one shadowing the other.
- One `this.ws` at the class level.
- One `manager.ws` which is "a stable alias for `this.ws` in callbacks" but is also sometimes used outside of callbacks. Though not always.

Since both `const ws` are constructed specifically in order to ultimately assign them to `manager.ws`, could we just always work with `manager.ws` directly? Could we also make sure we always refer to that property with `manager.ws`, rather than a mix of `this.ws` and `manager.ws`?

Since the `onWs*` functions are created (and presumably called?) as standalone functions, not as object properties, do we know that they end up being called in a context that rebinds `this`? If not, can we skip the `manager` alias and just use `this.ws` everywhere?

What do we gain by making QueryStreamManager a class? It seems we get some amount of confusion and complexity that obviously led to bugs, but I'm unclear on what we actually gain compared to having a `queryStreamManager` function that returns a simple JS object with the properties we need (and thus no `this` anywhere, aliased or not). That function could easily use a single `let ws`, like the preexisting `streamSubmit` did, avoiding opportunities for confusion.

Reply:
> Since the onWs* functions are created (and presumably called?) as standalone functions, not as object properties, do we know that they end up being called in a context that rebinds this? If not, can we skip the manager alias and just use this.ws everywhere?

It could be, I originally wrote the code with those being object properties, which is why I introduced the aliasing.

> Two const ws with nested scopes, one shadowing the other.

Avoiding the shadowing is probably good, thanks.

> One manager.ws which is "a stable alias for this.ws in callbacks" but is also sometimes used outside of callbacks. Though not always.

It should not be used outside of the callbacks.

> It seems we get some amount of confusion and complexity that obviously led to bugs, but I'm unclear on what we actually gain compared to having a queryStreamManager function that returns a simple JS object with the properties we need (and thus no this anywhere, aliased or not).

The usage of `this` does not require classes. Classes in JavaScript are syntactic sugar around functions, so I'm not sure that you can implement this logic without referring to the object's own properties. Can you make a short example of how you would get rid of the usages of `this`?

Reply:
> The usage of `this` does not require classes. Classes in JavaScript are syntactic sugar around functions, so I'm not sure that you can implement this logic without referring to the object's own properties. Can you make a short example of how you would get rid of the usages of `this`?

I'm aware that you _can_ use `this` without classes, but my point is that if you use classes you have to use it, whereas if you don't use classes you can pretend it doesn't exist, and that's usually my preferred approach to JavaScript. Specifically (ignoring TypeScript annotations):

```javascript
class QueryStreamsManager {
  private ws = undefined;
  private resetAllState() {
    // If I want to reset `ws` here, I _have to_ use `this`, and thus worry
    // about all the complexity it brings.
    this.ws = undefined;
  }
}
```

vs

```javascript
function queryStreamsManager() {
  let ws = undefined;
  let resetAllState = () => {
    // no `this` to be seen anywhere, so no need to worry about it being
    // dynamically rebound. I get lexical scope, which is better imo than
    // dynamic.
    ws = undefined;
  };
  return {resetAllState};
}
```

Reply:
To be clear, though, I'm not saying I want this PR to be blocked on this (non-trivial, given the size of this class) refactoring, even if I do manage to convince you it's worth doing.

Reply:
right i'll cleanup to remove `this` alias and instead use `manager` in all stream multiplexing related code to reduce confusion.

`QueryStreamsManager` class is useful to encapsulate/isolate all stream management related information into a class away from polluting the ledger code e.g mainly the stream queries registered so far, some state information (e.g  when stream was live) , plus additional data structures for lookup efficiency purposes.  The handle on the `ws` also allows for us to cleanup ws connections on queries change

Reply:
> `QueryStreamsManager` class is useful to encapsulate/isolate all stream management related information into a class away from polluting the ledger code e.g mainly the stream queries registered so far, some state information (e.g when stream was live) , plus additional data structures for lookup efficiency purposes. The handle on the `ws` also allows for us to cleanup ws connections on queries change

Which parts of that isolation would not work just as well as local, lexically-scoped variables within a function?

(These are not rhetorical questions, I'm genuinely curious to understand what I'm missing here.)

Reply:
so two main things that I think would not work(at least is not obvious to me) are
 - `streamSubmit` adds/registers/submits a new query/request and we store the state(total set) of queries to be used for creating new websocket connections(and reconnects) and reference it via `manager.queries` which IMO seems cleaner as the underlying impl for handling stream queries should not be a concern of the `Ledger` class.
 - on every new `streamSubmit` we also have to close the existing websocket and create a new one, dont think this works with lexically scoped variables but even if it did overall that be one gigantic method. Am not sure what standard javascript design practices are and if this is the norm.